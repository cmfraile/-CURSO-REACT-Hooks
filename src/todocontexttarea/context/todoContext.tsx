import { createContext } from "react";
import { todocrudch } from "../hooks/todocrudhook";

export const todoContext = createContext<any>({});
export const todoProvider = ({children}:any) => {
    return(<todoContext.Provider value={{...todocrudch()}}>{children}</todoContext.Provider>)
}