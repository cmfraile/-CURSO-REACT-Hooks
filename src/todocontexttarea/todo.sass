.col
    border: none !important
.litem
    display: flex
    justify-content: flex-end
    align-items: center
    margin: 2px
    padding: 2px
    max-width: 100%
    button
        max-height: 30px
        max-width: 30px
        min-width: 30px
        min-height: 30px
.margin
    margin-top: 50px !important