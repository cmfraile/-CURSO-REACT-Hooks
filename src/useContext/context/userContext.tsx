import { createContext } from "react";

{/*
export const userContext = createContext();
export const userProvider = ({children}) => {return (<userContext.Provider value={{hola:'mundo'}}>{children}</userContext.Provider>)}
*/}

export const userProvider = ({children}:any) => {
    const userContext = createContext({});
    return(<userContext.Provider value={{hola:'mundo'}}>{children}</userContext.Provider>)
}