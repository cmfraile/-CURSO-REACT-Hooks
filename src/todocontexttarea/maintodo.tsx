import { useContext, useEffect } from 'react';
import { useForma } from '../hooks/04-formwithcustomhook';
import { todoContext } from './context/todoContext';
import './todo.sass';

interface todobj {id:number,todo:string,done:boolean};

const TodoList = ({todos = []}:{todos:todobj[] | []}) => {
    console.log('render de lista');
    return(<><ul>{(todos) && todos
        .sort((a,b) => {
            const sorta = a.done ; const sortb = b.done;
            if(sorta == sortb){return 0}
            if(sorta){return 1} ; if(sortb){return -1}
            return 0;
        })
        .map(({id,todo,done}:todobj,i:number) => {
            const marginbool = ():boolean => {if(i == 0){return false} ; if(done && !todos[i-1].done){return true}else{return false}}
            return <TodoItem key={id} id={id} todo={todo} done={done} margin={marginbool()}/>})}</ul></>
        )};
const TodoItem = ({id,todo,done,margin}:any) => {
    const { setEdit } = useContext(todoContext);
    const { delTODO , endTODO } = useContext(todoContext);
    return(<><li    className={`litem ${(margin) ? 'margin' : ''}`} 
                    style={{textDecoration: (done) && 'line-through'}}
                    onDoubleClick={() => {setEdit({id,todo,done})}} 
                    key={id}>{todo}
    <button className='btn btn-danger' onClick={() => {delTODO({id,todo,done})}}></button>
    <button className='btn btn-primary' disabled={done} onClick={() => {endTODO({id,todo,done})}}></button>
    </li></>)
};
const TodoAdd = ({todos,onNewTodo,onEditTodo,onNuke}:{onNewTodo:any,onEditTodo:any,todos:todobj[],onNuke:any}) => {

    const {todo,onInputChange,onResetForm} = useForma({todo:''});
    const { edit , setEdit } = useContext<{edit:todobj,setEdit:any}>(todoContext);

    const validarmiddle = (todo:string):boolean => {
        const caso = todo.trim().toUpperCase();
        if(todo.trim().length <= 2){return false};
        if(todos.map(x => x.todo.toUpperCase()).includes(caso)){return false};
        return true
    }

    useEffect(() => {
        if(!edit){return}
        onInputChange({target:{name:'todo',value:edit.todo}});
    },[edit])
    
    return(
        <>
            <form
            onSubmit={(e) => {
                e.preventDefault();
                if(!validarmiddle(todo)){return}
                if(!edit){onNewTodo(todo.trim())}
                if(edit){onEditTodo({id:edit.id,todo,done:edit.done}) ; setEdit(undefined)}
                onResetForm();
            }}>
            <input
            className="form-control"
            type="text"
            name="todo"
            value={todo}
            placeholder={'aÃ±adir todo'}
            onChange={onInputChange}
            />
            </form>
            <button className="btn btn-danger" disabled={(todos.length == 0) && true} onClick={() => {if(confirm('esta seguro de mandar todo a la megamierda?')){onNuke() ; setEdit(undefined)}}}>NUKE</button>
            {(edit) && <code>{JSON.stringify(edit.todo)}</code>}
        </>
    )

}

export const TodoApp = () => {

    const { todos , addTODO , nukeTODO , editTODO } = useContext(todoContext);

    return (
        <>
        <h2>Todo app : {todos.length} tareas pendientes</h2>
        <div className="container">
            <div className="row">
                <div className="col"><TodoList todos={todos} /></div>
                <div className="col"><TodoAdd   onNewTodo={addTODO}
                                                onEditTodo={editTODO}
                                                onNuke={nukeTODO}
                                                todos={todos}/></div>
            </div>
        </div>
        </>
    )
}