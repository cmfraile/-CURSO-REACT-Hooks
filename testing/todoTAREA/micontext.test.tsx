import React from "react";
import { TodoApp } from '../../src/todocontexttarea/maintodo';
import { todoProvider as TDP } from "../../src/todocontexttarea/context/todoContext";
import { screen , render } from '@testing-library/react';

JSON.parse = jest.fn().mockImplementationOnce(() => []);

const main = () => {

    test('snapshot:',() => {
        expect(render(<TDP><TodoApp/></TDP>).container).toMatchSnapshot();
    });

    

}

describe('pruebas sobre mi TodoApp y el context',main);