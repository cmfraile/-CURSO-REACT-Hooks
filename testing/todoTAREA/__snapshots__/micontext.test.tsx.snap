// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas sobre mi TodoApp y el context snapshot: 1`] = `
<div>
  <h2>
    Todo app : 
    0
     tareas pendientes
  </h2>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <ul />
      </div>
      <div
        class="col"
      >
        <form>
          <input
            class="form-control"
            name="todo"
            placeholder="aÃ±adir todo"
            type="text"
            value=""
          />
        </form>
        <button
          class="btn btn-danger"
          disabled=""
        >
          NUKE
        </button>
      </div>
    </div>
  </div>
</div>
`;
